core/thegraph/querier.go

```diff
@@ -0,0 +1,46 @@
+package thegraph
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+type RetryQuerier struct {
+	GraphQLQuerier
+	PullInterval time.Duration
+	MaxRetries   int
+}
+
+var _ GraphQLQuerier = (*RetryQuerier)(nil)
+
+func NewRetryQuerier(q GraphQLQuerier, interval time.Duration, maxRetries int) *RetryQuerier {
+	return &RetryQuerier{
+		GraphQLQuerier: q,
+		PullInterval:   interval,
+		MaxRetries:     maxRetries,
+	}
+}
+
+func (q *RetryQuerier) Query(ctx context.Context, query any, variables map[string]any) error {
+
+	retryCount := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if retryCount > q.MaxRetries {
+				return errors.New("max retries exceeded")
+			}
+			retryCount++
+
+			err := q.GraphQLQuerier.Query(ctx, query, variables)
```

Comment:
Maybe in a future PR, but it should check `err` and only retry on transient errors (429/5xx). If the request itself is invalid, retries shouldn't be made.

Reply:
I'm not really sure how to tell in this case. Going to leave this for a future PR.